extension/parquet/include/parquet_crypto.hpp

```diff
@@ -71,17 +72,20 @@ class ParquetCrypto {
 
 public:
 	//! Decrypt and read a Thrift object from the transport protocol
-	static uint32_t Read(TBase &object, TProtocol &iprot, const string &key);
+	static uint32_t Read(TBase &object, TProtocol &iprot, const string &key, const shared_ptr<EncryptionState> &aes_p);
```

Comment:
If this is const anyways is there a reason for this to be a reference to a shared_ptr instead of an EncryptionState, or an optional_ptr if it can be NULL?

Reply:
It can never be NULL here, so `optional_ptr` does not work. If I change everything to `unique_ptr`, I ultimately get the following error in `parquet_crypto.cpp`:

```Call to implicitly-deleted copy constructor of 'unique_ptr<duckdb::EncryptionState>``` 

So, to me, `shared_ptr` seemed the best solution

Reply:
Can't we change it to take a `const EncryptionState &` then?

Reply:
I've tested this, but the when I try to avoid the pointers it calls the abstract class - which has no implementation. `EncryptionState` can either contain a `AESGCMStateSSL` or a `AESGCMStateMBEDTLS` state, but which one it is we can only know at runtime. So I think what happens if I use  `const EncryptionState &` is that the states are automatically casted to the abstract class, which calls then only contains 'empty' functions.

Reply:
The functions are virtual so that should work, that's also how the current `shared_ptr<EncryptionState> &` behaves, it's never cast to the concrete type right?
The only problem arises when you pass it by value instead of by references, then you get [object slicing](https://en.wikipedia.org/wiki/Object_slicing#:~:text=In%20C%2B%2B%20programming%2C%20object,%2C%20been%20%22sliced%20off%22)

Reply:
Hmm yes I think you're right. Let me look into it again! Thanks :)